File: src/10.graph/22.93.go
package graph

import (
	"strconv"
	"strings"
)

func RestoreIpAddresses(s string) []string {
	ans := []string{}
	onPath := []string{}
	n := len(s)
	var backtrack func(r int)

	backtrack = func(r int) {
		if r == n && len(onPath) == 4 {
			ans = append(ans, strings.Join(onPath, "."))
			return
		}

		for i := r; i < n; i++ {
			end := i + 1
			if len(onPath) == 3 {
				end = n
			}

			val := s[r:end]

			// 不能有前导 0
			if len(val) > 1 && val[0] == '0' {
				break
			}

			if t, _ := strconv.Atoi(val); t > 255 {
				break
			}

			onPath = append(onPath, val)
			backtrack(i + 1)
			onPath = onPath[:len(onPath)-1]
		}

	}

	backtrack(0)

	return ans
}

File: src/10.graph/22.93.ts
export function restoreIpAddresses(s: string): string[] {
    const ans: string[] = [];
    const onPath: string[] = [];
    const n = s.length;
    const backtrack = (r: number) => {
        // 结束条件, 当 r === n 时,说明走到了最后
        if (r === n && onPath.length === 4) {
            ans.push(onPath.join("."));
            return;
        }

        for (let i = r; i < n; i++) {
            const end = onPath.length === 3 ? n : i + 1;
            const val = s.slice(r, end);

            // val 不能有 前导 0 并且不能大于 255,
            if (Number(val) > 255 || (val.length >= 2 && val[0] === "0")) {
                break;
            }

            onPath.push(val);
            backtrack(i + 1);
            onPath.pop();
        }
    };

    backtrack(0);

    return ans;
}
